DIPs/DIP1xxx-ajie.md

```diff
@@ -0,0 +1,246 @@
+of `ref` meaning iteration by any method was officially considered a weakness.
+
+The reason this behaviour is a weakness is that the user cannot rely that `foreach(ref ...)`
+will iterate by reference. If a range with non-reference semantics is accidently
+passed to such a loop, it will be iterated by value, which is likely to be unexcepted.
+
+But in cases where the programmer does not modify the iteration variable, he/she does
+not care whether the iteration is done by reference or by value. Thus, when iterating
+by reference, the programmer should be able to choose whether the program will fallback
+to iteration by value when iteration by reference cannot be implemented, or just fail
+to compile.
+
+For function return values with a similar problem, there is already the differentation between
+`ref` and `auto ref` keywords[6]. For consistency, `auto ref` keyword for `foreach` loop
+variables can be considered the best canditate to become a way to iterate when fallback
+to iteration by value is desired, and `ref` when the fallback is not desired.
+
+Non-copyable `struct`s are an excellent aid when designing containers for RAII (Resource-
+Acquisition-Is-Initialization) principle. EMSI-Containers [2] are good examples of such
+non-copyable RAII containers. When one has many of those containers, it is natural to
+put them in a range. And with ranges comes the need to iterate.
+
+The iteration can be done in a general way with the current rules (see alternatives),
+or by always annotating/unannotating the iteration variable manually based on what is
+iterated over. But all of the former have their disadvantages, and the latter leads
+to needless changes in code when maintenance is done.
+
+## Description
+
+This DIP also proposes that, when encountering a `foreach` loop with `ref` keyword (example):
+```D
+foreach (ref loopVariable1; aggregate)
+{
+    loopVariable.doSomething();
+}
+```
+
+then, if and only if, elements of `aggregate` are rvalues [3], a deprectation
+message must be emitted, suggesting to annotate `loopVariable1` with
+`auto ref` keyword instead, explained below. The DIP does not propose a length
+for the deprectation period.
+
+This DIP also proposes, that when the compiler encounters a foreach statement:
+(example)
+
+```D
+foreach (auto ref loopVariable2; aggregate)
+{
+    loopVariable.doSomething();
+}
+```
+
+...then if, and only if, elements of `aggregate` are lvalues [3], the the
+above statement has exactly the same meaning as if it was written like this:
+
+```D
+foreach (ref loopVariable2; aggregate)
+{
+    loopVariable.doSomething();
+}
+```
+
+Otherwise, the statement is interpreted as:
+
+```D
+foreach (loopVariable2; aggregate)
+{
+    loopVariable.doSomething();
+}
+```
+
+`auto ref` should work in both generic and non-generic functions, and also when iterating
+over a tuple. It should also be allowed to be used in `static foreach`, but with no
```

Comment:
Which tuples? `std.typecons.Tuple`? Why allow `auto ref` in `static foreach`?

Reply:
Either tuple. I have to edit that.

Ss for allowing the feature in `static foreach`, I thought that as `auto ref` will iterate by value anyway if references are not available, it will still behave as excepted in static loops. Forbidding it there would be a bit like forbidding double imports.

Reply:
Regarding tuples, after looked at the language spec and testing a bit, I realized that the matter is a bit more complex than just "letting it work" when iterating over them.

I were more explicit on `AliasSeq` iteration now, which hopefully clarifies the situation. On the other hand, the DIP is now more complex. So if this part starts to need redesign after redesign, then I'm going to just ditch it.